File: 设计模式/程序/8-FilterPattern/FilterPattern(python)/Criteria.py
from Person import Person
import abc

# 基类
class Criteria(metaclass = abc.ABCMeta):
    @abc.abstractmethod
    def MeetCriteria(self, persons):
        pass

# 实现
class CriteriaMale(Criteria):
    def MeetCriteria(self, persons):
        malePerons = []
        for person in persons:
            if person.GetGender() == "Male":
                malePerons.append(person)
        return malePerons

class CriteriaFemale(Criteria):
    def MeetCriteria(self, persons):
        femalePersons = []
        for person in persons:
            if person.GetGender() == "Female":
                femalePersons.append(person)
        return femalePersons

class CriteriaSingle(Criteria):
    def __init__(self):
        pass
    def MeetCriteria(self, persons):
        singlePersons = []
        for person in persons:
            if person.GetMaritalStatus() == "Single":
                singlePersons.append(person)
        return singlePersons

# 与条件过滤
class AndCriteria(Criteria):
    def __init__(self, criteria1, criteria2):
        self.criteria1 = criteria1
        self.criteria2 = criteria2
    def MeetCriteria(self, persons):
        persons1 = self.criteria1.MeetCriteria(persons)
        persons2 = self.criteria2.MeetCriteria(persons1)
        return persons2

# 或条件过率
class OrCriteria(Criteria):
    def __init__(self, criteria1, criteria2):
        self.criteria1 = criteria1
        self.criteria2 = criteria2
    def MeetCriteria(self, persons):
        persons1 = self.criteria1.MeetCriteria(persons)
        persons2 = self.criteria2.MeetCriteria(persons)
        for person in persons1:
            if person not in persons2:
                persons2.append(person)
        return persons2

File: 设计模式/程序/8-FilterPattern/FilterPattern(C#)/Criteria.cs
using System;
using System.Collections.Generic;

namespace FilterPattern
{
    /* 接口 */
    public interface Criteria
    {
        List<Person> MeetCriteria(List<Person> persons);
    }

    /* 判断是否为男性 */
    public class CriteriaMale: Criteria
    {
        public List<Person> MeetCriteria(List<Person> persons)
        {
            List<Person> malePersons = new List<Person>();
            foreach (Person person in persons)
            {
                if (person.GetGender() == "Male")
                    malePersons.Add(person);
            }
            return malePersons;
        }
    }

    /* 判断是否为女性 */
    public class CriteriaFemale: Criteria
    {
        public List<Person> MeetCriteria(List<Person> persons)
        {
            List<Person> femalePersons = new List<Person>();
            foreach (Person person in persons)
            {
                if (person.GetGender() == "Female")
                    femalePersons.Add(person);
            }
            return femalePersons;
        }
    }

    /* 判断是否为女性 */
    public class CriteriaSingle: Criteria
    {
        public List<Person> MeetCriteria(List<Person> persons)
        {
            List<Person> singlePersons = new List<Person>();
            foreach (Person person in persons)
            {
                if (person.GetMaritalStatus() == "Single")
                    singlePersons.Add(person);
            }
            return singlePersons;
        }
    }

    /* 合并过滤 */
    public class AndCriteria: Criteria
    {
        // 构造函数
        public AndCriteria(Criteria criteria1, Criteria criteria2)
        {
            this.criteria1 = criteria1;
            this.criteria2 = criteria2;
        }
        // 实现接口
        public List<Person> MeetCriteria(List<Person> persons)
        {
            // 过滤两次
            List<Person> first_persons = this.criteria1.MeetCriteria(persons);
            return this.criteria2.MeetCriteria(first_persons);
        }

        // 两个条件
        private Criteria criteria1;
        private Criteria criteria2;
    }

    /* 或过滤 */
    public class OrCriteria: Criteria
    {
        // 构造函数
        public OrCriteria(Criteria criteria1, Criteria criteria2)
        {
            this.criteria1 = criteria1;
            this.criteria2 = criteria2;
        }
        // 实现接口
        public List<Person> MeetCriteria(List<Person> persons)
        {
            // 过滤两次
            List<Person> first_persons = this.criteria1.MeetCriteria(persons);
            List<Person> second_persons = this.criteria2.MeetCriteria(persons);
            // 把第二次过滤得到的链表放入第一个
            foreach (Person person in second_persons)
            {
                if (!first_persons.Contains(person))
                    first_persons.Add(person);
            }
            return first_persons;
        }

        // 两个条件
        private Criteria criteria1;
        private Criteria criteria2;   
    }

}